### ISSUE TITLE
Improve handling of Scoped Storage

### ISSUE THREAD

**Issue by zod** [abrensch/brouter#351]
As android introduced [scoped storage](https://developer.android.com/about/versions/11/privacy/storage) with android 10 (API 29) and enforces it with android 11 (API 30) brouter should improve it's handling. Scoped storage limits access on external storage as SD cards to only selected folders instead of the whole storage. The previous permission [WRITE_EXTERNAL_STORAGE](https://developer.android.com/reference/android/Manifest.permission#WRITE_EXTERNAL_STORAGE) no longer works when targeting API 30 (and perhaps it's even more restricted).

Each app has access to app specific folders without requesting any permissions. They can be done (and BRouter does) via [Context.getExternalMediaDirs()](https://developer.android.com/reference/android/content/Context#getExternalMediaDirs()) and [Context.getExternalFilesDirs()](https://developer.android.com/reference/android/content/Context#getExternalFilesDirs(java.lang.String)).

There are several [workarounds](https://developer.android.com/training/data-storage/use-cases#opt-out-scoped-storage) to opt out of the scoped storage behavior

- [requestLegacyExternalStorage](https://developer.android.com/reference/android/R.attr#requestLegacyExternalStorage) disables the scoped storage mode for the app. For apps targeting API 28 and lower it defaults to true, for apps targeting API 29 and higher it defaults to false. For apps targeting API 30 it is ignored and scoped storage is enforced.
- [preserveLegacyExternalStorage](https://developer.android.com/reference/android/R.attr#preserveLegacyExternalStorage) disables the scoped storage mode for the app if it's installed as an update from a previous version. This can be used to access existing files and provide a migration to a new folder which still can be accessed if scoped storage is enabled.

It seems like there are also issues with BRouter v.1.6.1 which targets API 28 as the instructions on [kurviger](https://docs.kurviger.de/start?id=app/offline_routing) suggest to use the API 10 version when using android 11. I'm not sure what exactly the issues are so if you have any details I'm eager to hear them.

I know that some things regarding storage have already been discussed in #312 but I think there's still room for improvement.

I suggest to set the `preserveLegacyExternalStorage` to allow easier updating and still accessing existing files. The `requestLegacyExternalStorage` could only be used if we target API 29 instead of the API 30 which we currently do. [Google Play's target API level requirement](https://developer.android.com/distribute/best-practices/develop/target-sdk) require targeting API 30 starting in November 2021.

Going forward the runtime permission handling should be improved as it crashed the app for me when it tried to request the permissions because of an [endless recursion](https://github.com/abrensch/brouter/blob/ac13b1fe34d49a9b20aae023f2c4f1fbba45600c/brouter-routing-app/src/api19/java/btools/routingapp/BRouterMainActivity.java#L48).

As I like task lists to keep an overview here is a first suggestion of the required steps
- [ ] Set `preserveLegacyExternalStorage` to allow upgrading users to still access their files on android 11
- [ ] Decide if target API should be lowered to API 29 to allow using `requestLegacyExternalStorage` before November 2021 and improve the permission handling
- [ ] Fix requesting permissions
- [ ] Improve migration from legacy storage (e.g. `/sdcard/emulated/0/brouter/`)

**Comment by devemux86** [comment]
> It seems like there are also issues with BRouter v.1.6.1 which targets API 28 as the instructions on kurviger suggest to use the API 10 version when using android 11. I'm not sure what exactly the issues are so if you have any details I'm eager to hear them.

This seems to be necessary when users want to change the BRouter profile files on Android 11+:
https://docs.kurviger.de/app/offline_routing

**Comment by afischerdev** [comment]
> I saw the migration code but I don't really understand the reason behind it. The access to /Android/data/ hasn't changed from BRouter's point of view. Is it just for other apps to access the profiles?

Yes, for edit profiles and store 'favourites.gpx' like on OsmAnd before.

> I'm having also some difficulties to understand why the "classic" routing mode (waypoint selection in BRouter app) isn't available since Android 6 since the access to the SD card wasn't limited, it's just required to request the permissions at runtime. From my understanding only the scoped storage feature with targetSdkVersion 30 would disable this mode on android 11+ devices.

I can't say why. There was no handling for WRITE_EXTERNAL_STORAGE.
But now this way is open so we could enable it. Let us wait for #312 and #354.
